osh/cmd_parse.py

```diff
@@ -1395,55 +1437,65 @@ def ParseCaseArm(self):
 
     self._NewlineOk()
 
-    return CaseArm(left_tok, pat_words, middle_tok, action_children, dsemi_tok)
+    return CaseArm(left_tok, pat.Words(pat_words), middle_tok, action_children, dsemi_tok)
 
-  def ParseYshCaseArm(self):
-    # type: () -> CaseArm
+  def ParseYshCaseArm(self, discriminant):
+    # type: (Id_t) -> Optional[Tuple[CaseArm, Id_t]]
```

Comment:
Although I said the "out param" was a smell, I feel like the Optional[Tuple[...]] is the same issue, i.e. the logic is a bit convoluted

The real "smell" is that _NewlineOkForYshCase has to peek into the lexer, because the WordParser is not able to expose enough info to the CommandParser

Reply:
You are right. If we handle the final `newline_ok` in `ParseYshCase` we can simplify this significantly.